### ISSUE TITLE
`:!someShellCommand` in nix repl

### ISSUE THREAD

**Issue by lf-** [NixOS/nix#6476]
**Is your feature request related to a problem? Please describe.**

It's frustrating to have to get out a separate terminal to run shell commands while I am exploring in the nix repl. For instance, I would like to be able to do `:!ls /nix/store/sdgfkjhlsgdflhk-somepackage/out`.

**Describe the solution you'd like**

`:!someCommand` will execute `someCommand` with the system shell synchronously, similar to how `!someCommand` works in `ipython`, `dc`, `vim`, `ed` and probably many others.

**Describe alternatives you've considered**

I could run nix repl in a split in my editor, but this interrupts my workflow as I would still have to switch panes to do shell commands.

**Additional context**

**Comment by aakropotkin** [comment]
Was just wish for this today because of using `nix repl` in `emacs` running in the project root vs where my file is.

Yeah I could handle it in `emacs`, but `:!ls` and `:!cd foo` would certainly be nice.

**Comment by thufschmitt** [comment]
Yes, I think that could make sense. I’d be happy to help with the implementation ;)

(not that `!cd` is a bit different because we can’t just shell out to on external command for that one. But maybe we could have it too)

**Comment by aakropotkin** [comment]
I implemented `ls` using `trace` and `readDir` for anyone who's interested. It supports globbing at the end of paths, and is easy to import into a REPL.

Certainly not a replacement for the features we discussed here, but some of you may find it useful.

https://github.com/aakropotkin/ak-nix/blob/main/lib/repl.nix#L35
